### PR TITLE
Model#updatePerson(...): Change parameter from Index to ReadOnlyPerson #468

### LINKED ISSUE #468: Change Model.updatePerson(int , ReadOnlyPerson) to Model.updatePerson(ReadOnlyPerson , ReadOnlyPerson)
The `Model` should not take in an `int index` as it exposes the internal detail of `Model` that it is using a `List` for its internal representation.

### PR DESCRIPTION
**Pull request by Zhiyuan-Amos**
Fixes #468

### CODE REVIEW

**Comment by CanIHasReview** [comment]
# v1

@Zhiyuan-Amos submitted v1 for review.

* [v1 1/1] [Model\#updatePerson\(\.\.\.\): Change parameter from Index to ReadOnlyPerson](https://github.com/se-edu/addressbook-level4/pull/471/commits/ad5b6378a1dd3565264eb2958c846ce44fb18a42)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/dfe3b25eb274119f2d0605fa266877597a3341c6...ad5b6378a1dd3565264eb2958c846ce44fb18a42))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/471/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 1]
```diff
@@ -73,9 +73,11 @@ public CommandResult execute() throws CommandException {
         Person editedPerson = createEditedPerson(personToEdit, editPersonDescriptor);
 
         try {
-            model.updatePerson(index, editedPerson);
+            model.updatePerson(personToEdit, editedPerson);
         } catch (UniquePersonList.DuplicatePersonException dpe) {
             throw new CommandException(MESSAGE_DUPLICATE_PERSON);
+        } catch (UniquePersonList.PersonNotFoundException pnfe) {
+            assert false : "The target person cannot be missing";
```
Comment (yamgent) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: `throw new AssertionError()`

File: src/main/java/seedu/address/model/AddressBook.java [hunk 2]
```diff
@@ -92,25 +93,25 @@ public void addPerson(ReadOnlyPerson p) throws UniquePersonList.DuplicatePersonE
...
 
     /**
-     * Updates the person in the list at position {@code index} with {@code editedReadOnlyPerson}.
+     * Replaces the given person {@code target} in the list with {@code editedReadOnlyPerson}.
      * {@code AddressBook}'s tag list will be updated with the tags of {@code editedReadOnlyPerson}.
      *
      * @throws DuplicatePersonException if updating the person's details causes the person to be equivalent to
      *      another existing person in the list.
-     * @throws IndexOutOfBoundsException if {@code index} < 0 or >= the size of the list.
+     * @throws PersonNotFoundException if no such person could be found in the list.
```
Comment (yamgent) [inline src/main/java/seedu/address/model/AddressBook.java]: `if {@code target} could not be found in the list.`

File: src/main/java/seedu/address/model/Model.java [hunk 3]
```diff
@@ -25,13 +25,14 @@
     void addPerson(ReadOnlyPerson person) throws UniquePersonList.DuplicatePersonException;
 
     /**
-     * Updates the person located at {@code index} with {@code editedPerson}.
+     * Replaces the given person {@code target} with {@code editedPerson}.
      *
      * @throws DuplicatePersonException if updating the person's details causes the person to be equivalent to
      *      another existing person in the list.
-     * @throws IndexOutOfBoundsException if {@code index} is out of range of the filtered list.
+     * @throws PersonNotFoundException if no such person could be found in the list.
```
Comment (yamgent) [inline src/main/java/seedu/address/model/Model.java]: same here

File: src/main/java/seedu/address/model/person/UniquePersonList.java [hunk 4]
```diff
@@ -45,15 +45,21 @@ public void add(ReadOnlyPerson toAdd) throws DuplicatePersonException {
     }
 
     /**
-     * Updates the person in the list at position {@code index} with {@code editedPerson}.
+     * Replaces the person {@code target} in the list with {@code editedPerson}.
      *
      * @throws DuplicatePersonException if updating the person's details causes the person to be equivalent to
      *      another existing person in the list.
-     * @throws IndexOutOfBoundsException if {@code index} < 0 or >= the size of the list.
+     * @throws PersonNotFoundException if no such person could be found in the list.
```
Comment (yamgent) [inline src/main/java/seedu/address/model/person/UniquePersonList.java]: same here

**Comment by CanIHasReview** [comment]
# v2

@Zhiyuan-Amos submitted v2 for review.

* [v2 1/1] [Model\#updatePerson\(\.\.\.\): Change parameter from Index to ReadOnlyPerson](https://github.com/se-edu/addressbook-level4/pull/471/commits/c2c7f61609aab6e049cf53fdd1d9fd6c217f52a3)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/dfe3b25eb274119f2d0605fa266877597a3341c6...c2c7f61609aab6e049cf53fdd1d9fd6c217f52a3)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/f5fcde6c2f03bba3538641dcc7a0fe7dfcb98a4e~1...f5fcde6c2f03bba3538641dcc7a0fe7dfcb98a4e))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/471/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/main/java/seedu/address/model/AddressBook.java [hunk 5]
```diff
@@ -92,25 +93,25 @@ public void addPerson(ReadOnlyPerson p) throws UniquePersonList.DuplicatePersonE
...
      *      another existing person in the list.
-     * @throws IndexOutOfBoundsException if {@code index} < 0 or >= the size of the list.
+     * @throws PersonNotFoundException if {@code target} could not be found in the list.
      *
      * @see #syncMasterTagListWith(Person)
      */
-    public void updatePerson(int index, ReadOnlyPerson editedReadOnlyPerson)
-            throws UniquePersonList.DuplicatePersonException {
+    public void updatePerson(ReadOnlyPerson target, ReadOnlyPerson editedReadOnlyPerson)
+            throws UniquePersonList.DuplicatePersonException, UniquePersonList.PersonNotFoundException {
```
Comment (damithc) [inline src/main/java/seedu/address/model/AddressBook.java]: Tangentially related: should the API exposes the class `UniquePersonList`?

Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/model/AddressBook.java]: As discussed with @yamgent, should we bring `DuplicatePersonException` and `PNFE` out into separate classes? :>

Comment (damithc) [inline src/main/java/seedu/address/model/AddressBook.java]: Looks like, given they are part of the API.

**Comment by pyokagan** [comment]
@Zhiyuan-Amos @damithc 
> Model exposes its internal representation by taking in an `int index`
> to update the person at the index.
>
> Model is a facade and should not expose its internal representation.

Model#getFilteredPersonList() already exposes the fact that it maintains an ordered collection of persons. So, why would taking an Index expose any more internal details than what is already exposed?

There's also the efficiency argument, of course, since we need to do a linear search now as well.

**Comment by Zhiyuan-Amos** [comment]
Yup that makes sense!

**Pull request merged**